import_models_to_verta/mlflow_databricks_bulk_import_example.ipynb

```diff
@@ -0,0 +1,341 @@
+    "            continue\n",
+    "\n",
+    "        # download artifacts\n",
+    "        try:\n",
+    "            outpath = mlflow.artifacts.download_artifacts(run_id=version.run_id)\n",
+    "        except Exception as err:\n",
+    "            model_messages[rm].append(f\"unable to download artifacts from {rm.name} - version run id; {version.run_id}\")\n",
+    "            outpath = None\n",
+    "\n",
+    "        if outpath:\n",
+    "            for file_name in os.listdir(outpath):\n",
+    "                if file_name not in protected_file_names:\n",
+    "                    verta_version.log_artifact(file_name, os.path.join(outpath, file_name))\n",
+    "                    model_messages[rm].append(f\"artifact logged in Verta: {file_name}\")\n",
+    "                else:\n",
+    "                    model_messages[rm].append(f\"cannot upload artifact {file_name} for {rm.name} as the file name conflicts with a protected file name in Verta\")\n",
+    "        else:\n",
+    "            model_messages[rm].append(f\"no artifacts found for {rm.name} - {version.version}\")\n",
+    "\n",
+    "        # set stage for version\n",
+    "        stage_error_message = f\"unable to set stage in Verta for {rm.name} - version: {version}, current_stage: {version.current_stage}\"\n",
+    "        try:\n",
+    "            if version.current_stage != \"None\":\n",
+    "                if version.current_stage == \"Staging\":\n",
+    "                    verta_version.change_stage(Staging())\n",
+    "                if version.current_stage == \"Production\":\n",
+    "                    verta_version.change_stage(Production())\n",
+    "                if version.current_stage == \"Archived\":\n",
+    "                    verta_version.change_stage(Archived())\n",
+    "            else:\n",
+    "                model_messages[rm].append(stage_error_message)\n",
+    "        except Exception as err:\n",
+    "            model_messages[rm].append(f\"{stage_error_message}, due to: {str(err)}\")\n",
+    "\n",
+    "        successful_model_imports.append(rm)\n"
+   ],
+   "metadata": {
+    "collapsed": false
+   }
+  },
+  {
+   "cell_type": "markdown",
+   "source": [
+    "## Print Results"
+   ],
+   "metadata": {
+    "collapsed": false
+   }
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "outputs": [],
+   "source": [
+    "print(\"IMPORTED SUCCESSFULLY:\\n\")\n",
+    "for m in successful_model_imports:\n",
+    "    print(f\"\\n{m.name}\\n---------------------\")\n",
+    "    for message in model_messages[m]:\n",
+    "        print(f\"  - {message}\")\n",
+    "\n",
+    "print(\"\\nFAILED TO IMPORT:\\n\")\n",
+    "for m in failed_model_imports:\n",
+    "    print(f\"\\n{m.name}\\n---------------------\")\n",
+    "    for message in model_messages[m]:\n",
+    "        print(f\"  - {message}\")"
+   ],
```

Comment:
Out of curiosity, was there a reason for printing this at the end, instead of printing it live during the `for` loop?

Reply:
I want to specifically call out models that succeeded in being imported but may have had some minor issue like an artifact that wasn't created, or the stage couldn't be set correctly, from those that failed entirely and resulted in nothing being created in Verta.  This format just allowed me to present the outcomes in a more coherent way than sequentially printed messages.